### DISCUSSION TITLE
Server on cloud VM and local 3D slicer

### DISCUSSION THREAD

**Discussion by St3V0Bay** [Project-MONAI/MONAILabel#558] [Q&A] [answered]
Dear all,

I ran into an error when the monailabel server is running on a cloud VM (in my case on GCP) and having 3D Slicer locally. When clicking the button "Next Sample" it could not load the image. I therefore assume that the images are forwarded by sending the file path from the server to the slicer (rather than the bytes themselves). This works when server and slicer are running on the same machine or in a LAN with relaxed read/write permissions.

![Screenshot 2021-12-15 at 08 48 24](https://user-images.githubusercontent.com/87860698/146148384-72a746c9-100b-461e-869f-6eb0c5b9e61f.png)

However, I want to make the case that the cloud2local scenario is highly important and should be supported. Large and distributed data consortia usually live in the cloud (the GPU compute as well as data storage), whereas medical experts needed for annotating use either their office laptops or special reader machine (as in clinical trials) - but of them don't have a GPU. So allowing a full outsourcing of the server (and its GPU compute) would be highly beneficial in terms of scaling.

@SachidanandAlle  - is the cloud2local scenario supported and I just made a mistake while setting it up? Or is this not yet implemented (if the latter is the case, I could help implementing, but would need some pointers).

**Comment by SachidanandAlle** [comment]
It is supported.. server can run anywhere and slicer will try to download the image on your local box.. as an optimization it tries to see if local path is accessible to avoid download (same host scenario)

You can see the slicer logs or server logs.. the next sample action reruns image id and url to download the image from server...

Or try this on your local box.. http://server:port/ and try download api... give the image id to download it.. the same is used by slicer..

**Reply by St3V0Bay** [reply]
you're fast :-). Highly appreciated. This is good information, so it *should* work already. 

And yes, for troubleshooting I will probably use the API directly (awesome that you have the Swagger UI). Let's see how far I can get.

**Comment by diazandr3s** [comment]
Hi @St3V0Bay,

Thanks for raising this issue.
Last week there was another colleague that had the same problem. It is because he was using HTTPS to connect to the server. 
Please try with HTTP. I mean, in Slicer you should use http://127.0.0.1:8001

I updated the README file ![http](https://user-images.githubusercontent.com/11991079/146166302-5e71cc0c-1cef-4e5e-8d51-540920addd18.png)

Please let us know how it goes.

Thanks again!

**Comment by St3V0Bay** [comment]
I could drill down to the respective line in MONAILabel.py. Basically a SampleData object should enable the download. I also see import SampleData but cannot find a corresponding package or lib. Could you provide a pointer? Thanks a lot
![image](https://user-images.githubusercontent.com/87860698/146201502-9c118752-553b-4a1e-b7da-6d6b5d96a639.png)

**Reply by St3V0Bay** [reply]
May be I need to add, that the download via Swagger does indeed work. So the API itself works.

**Reply by lassoan** [reply]
SampleData is a module in 3D Slicer. It is in `lib\Slicer-4.13\qt-scripted-modules\SampleData.py` in the Slicer install tree.

**Reply by St3V0Bay** [reply]
Thanks @lassoan. Your pointer helped a lot. Since on Mac OS any application is nicely wrapped as an app object (in Finder) I needed to use "right-click" trickery to actually reach the py libs.

**Comment by SachidanandAlle** [comment]
You can use python interface within 3d slicer

**Comment by St3V0Bay** [comment, marked as answer]
Dear community,

I could successfully connect a local 3D Slicer with MONAIlabel server running on Google Cloud Platform. One can basically use a regular notebook instance (e.g. https://cloud.google.com/vertex-ai-workbench#section-6) and tunnel the port to the local machine using this gcloud function. https://cloud.google.com/sdk/gcloud/reference/compute/start-iap-tunnel 

The data access and compute ressources (type and size of GPU) can easily be manage in GCP and the annotator only needs to open 3D Slicer w/ MONAILabel and is ready to go. Best thing since sliced bread :-)

**Reply by diazandr3s** [reply]
This is great! Thanks for letting us know :)

**Reply by PaWeRe** [reply]
Thanks for your advice on how to run MONAILabel server on GCP @St3V0Bay and connect to 3DSlicer! Does this mean there's currently no way to run MONAILabel server locally on mac os @SachidanandAlle? 

Thanks for your time in advance! :)

**Reply by SachidanandAlle** [reply]
Limited support for CPU.. some people have tried basic (infer) on CPU.. but user experience may not that great.. you can give a try.. unless some operations on in the pre trained/network is not gpu specific it should work..

**Reply by PaWeRe** [reply]
Thanks for the quick reply @SachidanandAlle, I'll give it a try!

**Reply by diazandr3s** [reply]
Hi @PaWeRe, 

Just in case you're interested, there is a MONAI Label workshop at GTC this year: https://www.nvidia.com/gtc/session-catalog/?search=DLIT2098&tab.scheduledorondemand=1583520458947001NJiE#/

Limited places

**Reply by PaWeRe** [reply]
Sounds interesting @diazandr3s, is the workshop for free? :)

**Reply by diazandr3s** [reply]
Yes, it is ... limited places, though

Today we're also hosting a virtual developer meetup for researchers and developers of the medical imaging community ... Come and join us!

https://events.rainfocus.com/widget/nvidia/gtcspring2022/16443900066980018eO0?ncid=so-nvsh-474528-vt12

![MONAI](https://user-images.githubusercontent.com/11991079/159257092-1e8efb61-81d2-4a70-b681-3d9938599412.png)

**Reply by PaWeRe** [reply]
Was not able to make it to the developer meetup on Monday, but will join todays workshop :)
